marimo/_server/sessions.py

```diff
@@ -212,49 +220,110 @@ def create(
+        # Reads from the kernel connection and distributes the
+        # messages to each subscriber.
+        self.message_distributor = Distributor[KernelMessage](
+            self.kernel_manager.kernel_connection
+        )
+        self.message_distributor.add_consumer(
+            lambda msg: self.session_view.add_raw_operation(msg[1])
         )
+        self.connect_consumer(session_consumer)
+        self.message_distributor.start()
+
+    def _check_alive(self) -> None:
+        if not self.kernel_manager.is_alive():
+            LOGGER.debug("Closing session because kernel died")
+            self.close()
+            print()
+            print_tabbed("\033[31mThe Python kernel died unexpectedly.\033[0m")
+            print()
+            sys.exit()
 
     def try_interrupt(self) -> None:
         self.kernel_manager.interrupt_kernel()
 
     def put_request(self, request: requests.Request) -> None:
         self._queue_manager.control_queue.put(request)
+        self.session_view.add_request(request)
 
     def put_input(self, text: str) -> None:
         self._queue_manager.input_queue.put(text)
+        self.session_view.add_stdin(text)
+
+    def disconnect_consumer(self) -> None:
+        """Stop the session consumer but keep the kernel running"""
+        assert (
+            self.session_consumer is not None
+        ), "Expecting a session consumer to pause"
+        LOGGER.debug("Disconnecting session consumer")
+        self.session_consumer.on_stop()
+        self.unsubscribe_consumer()
+        self.session_consumer = None
+
+    def connect_consumer(self, session_consumer: SessionConsumer) -> None:
+        """Connect or resume the session with a new consumer"""
+        assert (
+            self.session_consumer is None
+        ), "Expecting no existing session consumer"
+
+        self.session_consumer = session_consumer
+
+        subscribe = self.session_consumer.on_start(self._check_alive)
+        self.unsubscribe_consumer = self.message_distributor.add_consumer(
+            subscribe
+        )
+
+    def get_current_state(self) -> SessionView:
+        return self.session_view
+
+    def connection_state(self) -> ConnectionState:
+        if self.session_consumer is None:
+            return ConnectionState.ORPHANED
+        return self.session_consumer.connection_state()
+
+    async def write_operation(self, operation: MessageOperation) -> None:
+        self.session_view.add_operation(operation)
+        if self.session_consumer is not None:
+            await self.session_consumer.write_operation(operation)
 
     def close(self) -> None:
-        self.session_handler.on_stop(self.kernel_manager.kernel_connection)
+        # Could be no consumer if we already disconnect, but the session
+        # is running in the background
+        if self.session_consumer is not None:
+            self.session_consumer.on_stop()
         self.kernel_manager.close_kernel()
+        self.message_distributor.stop()
```

Comment:
One of these two methods needs to remove the reader on the multiprocessing connection from the event loop:

```python
asyncio.get_running_loop().remove_reader(self.kernel_connection.fileno())
```

Failure to remove the reader causes the server to hang on exit on Windows (I was only able to free up the terminal by using the Windows Task Manager to forcibly kill the Python process).

Additionally, the reader must be removed _before_ the kernel connection is closed (which currently happens in `kernel_manager.close_kernel()`, otherwise an `OSError` will be thrown

Reply:
good call. i originally did remove the reader, but removed that recently. i added it back